### ISSUE TITLE
SPZeroDownTimePatching: New resource for enabling ZDP with DSC in SP 2016 and 2019

### ISSUE THREAD

**Issue by andikrueger** [PowerShell/SharePointDsc#1097]
Following up on #1096 about ZDP there would be the need to create a new resource to handle Side-by-Side in this scenario. The new resource could look like this:

Enabling SideBySide could look like this:

```powershell
SPZeroDownTimePatching EnableSideBySide {
    EnableSideBySide = $true
}
```

Enabling SideBySide could look like this:

```powershell
SPZeroDownTimePatching EnableSideBySide {
    EnableSideBySide = $false
}
```

All information needed to set the web application property `WebSerivce.SideBySide` can be obtained within this resource, by querying the farm or using a part of the new methods introduced with the refactoring of `SPProductUpdate`

**Comment by ykuijs** [comment]
See my comment in the other issue. I don't think this resource is needed, since the two examples will be in constant conflict and the server will never be in a compliant state.

**Comment by andikrueger** [comment]
I think, there is a chance to  resolve this potential conflict. 

Both examples need to return true (test-method) for the following case:

Local Version obtained from the SPProductUpdate Method equals the Farm Version and SideBySide is enabled and the SideBySideToken equals the Farm Version 

False for 
- first example: 
— SideBySide is not enabled
— Version mismatch

- second example:
— SideBySide is not enabled 
— Version mismatch 

Basically both would return the same status - the set method would differ though. I’m “Enable”  the set method would pass the current Version to SharePoint. In “Upgrade” it would set the installed Version.

**Comment by mrpullen** [comment]
wouldn't this work?

```
SPZeroDownTimePatching EnableSideBySide {
    MinimumVersion = "16.0.4888.1000" 
    Enabled = "Present"
}

... handle cross node dependencies / order
SPProductUpdate LatestPatch {}
... wait for all product updates to complete / cross node dependencies / ordering.
SPConfigWizard ConfigWizard {}
... handle any reboot requirements / ensure another WFE is available to serve requests.

SPZeroDownTimePatching EnableSideBySide {
    MinimumVersion = "16.0.4900.1000" 
    Enabled = "Present"
}
```

1. Enabled would control your sideBySide being enabled or disabled.
2. MinimumVersion would handle version matching by the following rule

The idea would be if the Test-Resource would return true if local version on a specific node was equal to or greater than the version specified. Could also just include a "Version" tag to force set to a specific version (perhaps as part of an install scriptj).

**Comment by ykuijs** [comment]
The suggestion for the SPZeroDownTimePatching resource would work indeed. The MinimumVersion needs to be Key parameter in this case. And out of interest, what would the Enabled parameter do in this case?

**Comment by andikrueger** [comment]
Using the parameter `MinimumVersion` to handle the potential conflict of my suggestions is a very good idea.

@ykuijs For my understanding the parameter `Enabled` will control the SideBySide Setting of the web application: https://docs.microsoft.com/en-us/previous-versions/office/sharepoint-server/dn778174(v=office.15)?redirectedfrom=MSDN
